### PR TITLE
[SHELL32] Implement shell change notification

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

This PR is a retrial of #2411 but it uses `WorkerW` windows.

- Revert @AmineKhaldi's work at `wine/changenotify.c`.
- Rewrite `SHChangeNotify`, `SHChangeNotifyRegister`, `SHChangeNotify_Lock`, `SHChangeNotify_Unlock` and `SHChangeNotifyDeregister` functions.
- Fix `SHSimpleIDListFromPathA/W` functions.

JIRA issue: [CORE-13950](https://jira.reactos.org/browse/CORE-13950)

### CODE REVIEW

**Comment by katahiromz** [comment]
Not working at all.
not-working-at-all

**Comment by katahiromz** [comment]
```txt
trace:(dll/win32/shell32/CChangeNotify.cpp:91) DoCreateNotifWindow entered
trace:(dll/win32/shell32/CChangeNotify.cpp:97) DoCreateNotifWindow leaved
trace:(dll/win32/shell32/CChangeNotify.cpp:72) ChangeNotifThreadFunc entered
(win32ss/user/ntuser/message.c:1440) err: Window 00020130 (BC651F40) (pti B0A4AA38) is hung!
```

**Comment by katahiromz** [comment]
analyzing

**Comment by katahiromz** [comment]
`CChangeNotify::m_items` must be passed beyond process boundary...

**Comment by katahiromz** [comment]
VirtualBox_ReactOS_14_03_2020_15_51_05
`HWND 000200F8` is old delivery worker.
`HWND 000200F6` is new delivery worker.
The delivery worker windows are correctly created.

**Comment by katahiromz** [comment]
Analyzing the log...
https://jira.reactos.org/secure/attachment/56242/log.txt

**Comment by katahiromz** [comment]
successful
Testcase became successful.

File: dll/win32/shell32/shelldesktop/CChangeNotify.cpp [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+        memcpy((LPBYTE)pTicket + ibOffset1, pidl1, cbPidl1);
+    if (pidl2)
+        memcpy((LPBYTE)pTicket + ibOffset2, pidl2, cbPidl2);
+
+    SHUnlockShared(pTicket);
+    return hTicket;
+}
+
+EXTERN_C LPHANDBAG
+DoGetHandBagFromTicket(HANDLE hTicket, DWORD dwOwnerPID)
```
Comment (ThFabba) [inline dll/win32/shell32/shelldesktop/CChangeNotify.cpp:229]: Handbag?!

File: dll/win32/shell32/shelldesktop/CChangeNotify.cpp [hunk 2]
```diff
@@ -0,0 +1,581 @@
...
+        LRESULT lResult = 0;
+        pThis->ProcessWindowMessage(hwnd, uMsg, wParam, lParam, lResult, 0);
+        return lResult;
+    }
+    return 0;
+}
+
+BOOL CWorker::CreateWorker(HWND hwndParent, DWORD dwExStyle, DWORD dwStyle)
+{
+    m_hWnd = SHCreateWorkerWindowW(WindowProc, hwndParent, dwExStyle, dwStyle,
```
Comment (ThFabba) [inline dll/win32/shell32/shelldesktop/CChangeNotify.cpp:294]: assert that `m_hWnd` is null before overwriting it?

File: dll/win32/shell32/shelldesktop/CChangeNotify.cpp [hunk 3]
```diff
@@ -0,0 +1,581 @@
...
+
+CChangeNotify::CChangeNotify()
+    : m_nNextRegID(INVALID_REG_ID)
+    , m_pimpl(new CChangeNotifyImpl)
+{
+}
+
+CChangeNotify::~CChangeNotify()
+{
+    delete m_pimpl;
```
Comment (ThFabba) [inline dll/win32/shell32/shelldesktop/CChangeNotify.cpp:362]: CComHeapPtr?

File: dll/win32/shell32/shelldesktop/CChangeNotify.cpp [hunk 4]
```diff
@@ -0,0 +1,581 @@
...
+}
+
+void CChangeNotify::clear()
+{
+    m_hWnd = NULL;
```
Comment (ThFabba) [inline dll/win32/shell32/shelldesktop/CChangeNotify.cpp:367]: Who destroys the window, if there is one?

File: dll/win32/shell32/wine/changenotify.c [hunk 5]
```diff
@@ -165,30 +153,26 @@ static void DeleteNode(LPNOTIFICATIONLIST item)
...
 }
 
 void FreeChangeNotifications(void)
 {
+#ifdef __REACTOS__
+    HWND hwndWorker;
+    EnterCriticalSection(&SHELL32_ChangenotifyCS);
+    hwndWorker = DoGetNewDeliveryWorker();
+    SendMessageW(hwndWorker, WM_NOTIF_REMOVEBYPID, GetCurrentProcessId(), 0);
+    LeaveCriticalSection(&SHELL32_ChangenotifyCS);
```
Comment (ThFabba) [inline dll/win32/shell32/wine/changenotify.c:173]: What other code is this protecting against and what guarantees that said code won't try to access the critical section after it's been deleted?
(lock-unlock-delete is usually a sign that the synchronization design is not safe)

File: dll/win32/shell32/wine/changenotify.c [hunk 6]
```diff
@@ -333,12 +369,96 @@ static BOOL should_notify( LPCITEMIDLIST changed, LPCITEMIDLIST watched, BOOL su
...
+            }
+            pidl1 = (LPITEMIDLIST)dwItem1;
+            pidl2 = (LPITEMIDLIST)dwItem2;
+            break;
+
+        case SHCNF_PATHA:
+            psz1 = psz2 = NULL;
+            if (dwItem1)
+            {
+                SHAnsiToUnicode((LPCSTR)dwItem1, szPath1, ARRAYSIZE(szPath1));
```
Comment (ThFabba) [inline dll/win32/shell32/wine/changenotify.c:411]: Cast is not needed here (and below) since this is C.

File: sdk/include/psdk/shlwapi.h [hunk 7]
```diff
@@ -1892,6 +1892,13 @@ SHLockShared(
     _In_  DWORD dwProcessId
     );
 
+PVOID
+WINAPI
+SHLockSharedEx(
```
Comment (ThFabba) [inline sdk/include/psdk/shlwapi.h:1897]: This doesn't belong here. undocshell.h perhaps?

File: dll/win32/shell32/shelldesktop/CChangeNotify.h [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+    virtual ~CChangeNotify();
+
+    operator HWND()
+    {
+        return m_hWnd;
+    }
+
+    void SetHWND(HWND hwnd)
+    {
+        m_hWnd = hwnd;
```
Comment (ThFabba) [inline dll/win32/shell32/shelldesktop/CChangeNotify.h:125]: What happens to the old window, if there was one?

File: dll/win32/shell32/shelldesktop/CChangeNotify.cpp [hunk 9]
```diff
@@ -0,0 +1,581 @@
...
+{
+    return m_pimpl->RemoveItem(nRegID, dwOwnerPID);
+}
+
+void CChangeNotify::RemoveItemsByProcess(DWORD dwOwnerPID, DWORD dwUserPID)
+{
+    m_pimpl->RemoveItemsByProcess(dwOwnerPID, dwUserPID);
+}
+
+LRESULT CChangeNotify::OnBang(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
```
Comment (ThFabba) [inline dll/win32/shell32/shelldesktop/CChangeNotify.cpp:387]: What's a bang?

File: dll/win32/shell32/shelldesktop/CChangeNotify.cpp [hunk 10]
```diff
@@ -0,0 +1,581 @@
...
+    return 0;
+}
+
+EXTERN_C HWND
+DoHireOldDeliveryWorker(HWND hwnd, UINT wMsg)
+{
+    LPOLDDELIVERYWORKER pWorker;
+    HWND hwndOldWorker;
+
+    pWorker = (LPOLDDELIVERYWORKER)LocalAlloc(LMEM_FIXED, sizeof(OLDDELIVERYWORKER));
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/shelldesktop/CChangeNotify.cpp]: OK.

File: dll/win32/shell32/shelldesktop/CChangeNotify.cpp [hunk 11]
```diff
@@ -0,0 +1,581 @@
...
+    }
+
+    ibOffset2 = 0;
+    if (pidl2)
+    {
+        cbPidl2 = ILGetSize(pidl2);
+        ibOffset2 = DWORD_ALIGNMENT(ibOffset1 + cbPidl1);
+    }
+
+    cbSize = ibOffset2 + cbPidl2;
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/shelldesktop/CChangeNotify.cpp:199]: Initialized.

File: dll/win32/shell32/shelldesktop/CChangeNotify.cpp [hunk 12]
```diff
@@ -0,0 +1,581 @@
...
+    {
+        cbPidl1 = ILGetSize(pidl1);
+        ibOffset1 = DWORD_ALIGNMENT(sizeof(DELITICKET));
+    }
+
+    ibOffset2 = 0;
+    if (pidl2)
+    {
+        cbPidl2 = ILGetSize(pidl2);
+        ibOffset2 = DWORD_ALIGNMENT(ibOffset1 + cbPidl1);
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/shelldesktop/CChangeNotify.cpp:196]: Initialized.

File: dll/win32/shell32/shelldesktop/CChangeNotify.cpp (hunk 2 above)
Comment (katahiromz) [reply dll/win32/shell32/shelldesktop/CChangeNotify.cpp:296]: Added assertion.

File: dll/win32/shell32/wine/changenotify.c (hunk 6 above)
Comment (katahiromz) [reply dll/win32/shell32/wine/changenotify.c]: Deleted.

File: dll/win32/shell32/shelldesktop/CChangeNotify.h (hunk 8 above)
Comment (katahiromz) [reply dll/win32/shell32/shelldesktop/CChangeNotify.h]: Deleted `SetHWND`.

File: dll/win32/shell32/shelldesktop/CChangeNotify.cpp (hunk 9 above)
Comment (katahiromz) [reply dll/win32/shell32/shelldesktop/CChangeNotify.cpp]: Renamed as `OnReg`.

File: dll/win32/shell32/shelldesktop/CChangeNotify.cpp (hunk 3 above)
Comment (katahiromz) [reply dll/win32/shell32/shelldesktop/CChangeNotify.cpp:364]: `m_pimpl` is not a COM pointer.

File: dll/win32/shell32/shelldesktop/CChangeNotify.cpp (hunk 4 above)
Comment (katahiromz) [reply dll/win32/shell32/shelldesktop/CChangeNotify.cpp]: `CChangeNotify::clear()` is deleted.

File: sdk/include/psdk/shlwapi.h (hunk 7 above)
Comment (katahiromz) [reply sdk/include/psdk/shlwapi.h]: Moved.

File: dll/win32/shell32/shelldesktop/CChangeNotify.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shell32/shelldesktop/CChangeNotify.cpp:230]: Can you suggest another name?

File: dll/win32/shell32/wine/changenotify.c (hunk 5 above)
Comment (katahiromz) [reply dll/win32/shell32/wine/changenotify.c]: I deleted some critical sections. I don't know sync well...

File: dll/win32/shell32/shelldesktop/CChangeNotify.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shell32/shelldesktop/CChangeNotify.cpp]: `s/HandBag/Handbag/`.

**Comment by yagoulas** [comment]
I like the general idea, I'll have a closer look today. I think there are some tiny bits that can be improved. This means that please don't hurry to merge it (yet).

File: dll/win32/shell32/shelldesktop/CChangeNotify.cpp (hunk 1 above)
Comment (ThFabba) [reply dll/win32/shell32/shelldesktop/CChangeNotify.cpp]: I'm not sure what a "Handbag" represents. "registration entry" perhaps?

Comment (katahiromz) [reply dll/win32/shell32/shelldesktop/CChangeNotify.cpp]: `SHChangeNotification_Lock` returns a local memory block, that is a handbag.

File: dll/win32/shell32/shelldesktop/CChangeNotify.h [hunk 13]
```diff
@@ -132,14 +132,14 @@ class CChangeNotify : public CWorker
...
-    LRESULT OnBang(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled);
+    LRESULT OnReg(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled);
     LRESULT OnUnReg(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled);
     LRESULT OnDelivery(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled);
     LRESULT OnSuspendResume(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled);
     LRESULT OnRemoveByPID(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled);
 
     BEGIN_MSG_MAP(CChangeNotify)
-        MESSAGE_HANDLER(WM_NOTIF_BANG, OnBang)
+        MESSAGE_HANDLER(WM_NOTIF_BANG, OnReg)
```
Comment (HBelusca) [inline dll/win32/shell32/shelldesktop/CChangeNotify.h:138]: Then, should this message be called `WM_NOTIF_REG` as well?

Comment (katahiromz) [reply dll/win32/shell32/shelldesktop/CChangeNotify.h]: Renamed.

**Comment by katahiromz** [comment]
There are two problems:
- New folder has wrong icon.
- Explorer won't respond `SHCNE_RMDIR`.
problem

**Comment by katahiromz** [comment]
VirtualBox_ReactOS_06_04_2020_10_54_56
Explorer is correctly receiving `SHCNE_RMDIR`. Thus, that is the problem of `ILIsParentOrSpecialParent` function.

**Comment by katahiromz** [comment]
Ready to go!

**Comment by katahiromz** [comment]
When will the old workers be destroyed?

**Comment by katahiromz** [comment]
Good luck, @yagoulas.

**Comment by yagoulas** [comment]
Ah I see, it was so urgent to merge this. My overall sense is that this crap is more or less correct. Yet it is unreadable. Can you at least try to improve it? 


There is a CWorker, a CChangeNotifyImpl, a CChangeNotify and there is wine's NOTIFICATIONLIST. Plus there is wine's code that is totally disabled. Since we no longer depend on wine's code we could just delete changenotify.c and move the code in CChangeNotify.cpp for starters

**Comment by yagoulas** [comment]
@katahiromz 
Ah I see, it was so urgent to merge this. My overall sense is that this crap is more or less correct. Yet it is unreadable. Can you at least try to improve it?

There is a CWorker, a CChangeNotifyImpl, a CChangeNotify and there is wine's NOTIFICATIONLIST. Plus there is wine's code that is totally disabled. Since we no longer depend on wine's code we could just delete changenotify.c and move the code in CChangeNotify.cpp for starters

Also, good luck to you as well

**Comment by katahiromz** [comment]
Ok.
■

**Comment by yagoulas** [comment]
The problem is that since it is unreadable I can't be sure about if some details are correct or not.

**Comment by katahiromz** [comment]
Ok, I will insert human readable comments...
shelldesktop is static library. it needs restructuring.

**Comment by yagoulas** [comment]
shelldesktop does have some exports. It was designed to be like that. Since changenotify is indeed somewhat tangled with desktop it is fine to keep it there. CDesktopBrowser.cpp also has a couple of exports and this is by design. CChangeNotify is fine there.

**Pull request merged**